### PR TITLE
New api for Marketplace repository

### PR DESCRIPTION
**Pull request by AlexanderPrendota**

### CODE REVIEW

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/internal/utils/RequestHelper.kt [hunk 1]
```diff
@@ -18,7 +18,12 @@ internal fun <T> executeAndParseBody(callable: Call<T>): T? {
   if (response.isSuccessful) {
     return response.body()
   }
-  val message = response.errorBody()?.string() ?: response.message() ?: Messages.getMessage("failed.request.status.code", response.code())
+  val message = listOf(
```
Comment (AlexanderPrendota) [inline rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/internal/utils/RequestHelper.kt]: @serejke I made your function smarter :smekalochka:

**Comment by kesarevs** [comment]
Sometimes you're using only `@param`, sometimes only `@return`, sometimes there is only a description. Could you please unify this. Also there is no information about exceptions.

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 2]
```diff
@@ -1,21 +1,144 @@
...
+  /**
+   * Plugin info by [xmlId] & [ProductFamily].
+   * @param xmlId - id from plugin.xml file. Example: "org.jetbrains.kotlin"
+   */
+  fun getPluginByXmlId(xmlId: String, family: ProductFamily = ProductFamily.INTELLIJ): PluginBean?
 
-  fun download(pluginXmlId: String, version: String, channel: String? = null, targetPath: File): File?
+  /**
+   * Plugin info by [id].
+   * Supported for all [ProductFamily].
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: Supported for all [ProductFamily].

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 3]
```diff
@@ -1,21 +1,144 @@
...
+interface PluginUpdateManager {
+  /**
+   * Get list of plugin updates by plugin [xmlId].
+   * @param version - version of the plugin.
+   * @return Normally method returns only ONE update, but several very OLD plugins have several update.
+   */
+  fun getUpdatesByVersionAndFamily(xmlId: String, version: String, family: ProductFamily = ProductFamily.INTELLIJ): List<PluginUpdateBean>
+
+  /**
+   * Get plugin update by [id]. If needed to get a lot of plugin updates recommended using [getIntellijUpdateMetadata].
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: s3

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 4]
```diff
@@ -1,21 +1,144 @@
...
+   * Getting plugin update metadata.
+   * Use for getting a big list of plugin updates.
+   * Supported for [ProductFamily.INTELLIJ].
+   */
+  fun getIntellijUpdateMetadata(pluginId: Int, updateId: Int): IntellijUpdateMetadata?
+}
+
+interface PluginDownloader {
+  /**
+   * Download [ProductFamily.INTELLIJ] plugin by plugin XML id.
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: yes

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 5]
```diff
@@ -1,21 +1,144 @@
...
+   */
+  fun download(xmlId: String, version: String, targetPath: File, channel: String? = null): File?
+
+  /**
+   * Download latest compatible update for plugin [ProductFamily.INTELLIJ] by IDE Version.
+   * @param xmlId    - plugin XML id.
+   * @param ideBuild - IDE version. Example: "IC-145.184"
+   * @param channel  - plugin channel. Default value is "stable" plugin channel.
+   */
+  fun downloadLatestCompatiblePlugin(xmlId: String, ideBuild: String, targetPath: File, channel: String? = null): File?
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: add more description pls

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 6]
```diff
@@ -1,21 +1,144 @@
...
+}
+
+interface PluginUploader {
+  /**
+   * Upload plugin by ID into specific channel.
+   * IMPORTANT: PLUGIN NOTES WILL BE IGNORED FOR IDES BASED ON INTELLIJ PLATFORM ([ProductFamily.INTELLIJ]).
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: thx

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 7]
```diff
@@ -1,21 +1,144 @@
...
+   * @param channel   - plugin channel. Default value is "stable" plugin channel.
+   * @param notes     - plugin update notes.
+   */
+  fun uploadPlugin(id: Int, file: File, channel: String? = null, notes: String? = null)
+
+  /**
+   * Upload plugin by Xml id into specific channel.
+   * IMPORTANT: PLUGIN NOTES WILL BE IGNORED FOR IDES BASED ON INTELLIJ PLATFORM ([ProductFamily.INTELLIJ]).
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: ok

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 8]
```diff
@@ -1,21 +1,144 @@
...
-  fun downloadCompatiblePlugin(pluginXmlId: String, ideBuild: String, channel: String? = null, targetPath: File): File?
+  /**
+   * List of plugin authors.
+   */
+  fun getPluginDevelopers(id: Int): List<PluginUserBean>
 
-  fun uploadPlugin(pluginId: Int, file: File, channel: String? = null, notes: String? = null)
+  /**
+   * List of plugins channels.
+   * Example: "stable", "", "EAP" and etc.
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: make sense.

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (kesarevs) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: downloadLatestCompatible**Plugin** vs Download latest compatible **update**

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 9]
```diff
@@ -1,21 +1,144 @@
...
+  val pluginUpdateManager: PluginUpdateManager
+  val downloader: PluginDownloader
+  val uploader: PluginUploader
+}
 
-  fun pluginInfo(family: String, pluginXmlId: String): PluginInfoBean?
+interface PluginManager {
+  /**
+   * Plugin info by [xmlId] & [ProductFamily].
+   * @param xmlId - id from plugin.xml file. Example: "org.jetbrains.kotlin"
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: Year. make sence

[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: year, make sense

File: README.md [hunk 10]
```diff
@@ -1,14 +1,46 @@
-# plugin-repository-rest-client [![JetBrains team project](https://jb.gg/badges/team.svg)](https://confluence.jetbrains.com/display/ALL/JetBrains+on+GitHub)
+# Plugin Repository Rest Client [![JetBrains team project](https://jb.gg/badges/team.svg)](https://confluence.jetbrains.com/display/ALL/JetBrains+on+GitHub)
 
 Client and command line interface for [JetBrains Plugin Repository](https://plugins.jetbrains.com/).
 
-The code snippet below will upload a new version of a plugin to the plugins repository.
+# Examples
+The code snippet below will init the instance of the JetBrains Marketplace repository.
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply README.md]: what kind of info?

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 11]
```diff
@@ -1,21 +1,144 @@
...
-  fun uploadPlugin(pluginId: Int, file: File, channel: String? = null, notes: String? = null)
+  /**
+   * List of plugins channels.
+   * Example: "stable", "", "EAP" and etc.
+   */
+  fun getPluginChannels(id: Int): List<String>
 
-  fun uploadPlugin(pluginXmlId: String, file: File, channel: String? = null, notes: String? = null)
+  /**
+   * List of plugin supported [ProductEnum].
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: thx

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 12]
```diff
@@ -1,21 +1,144 @@
...
+  /**
+   * List of plugin dependencies. Examples: "com.intellij.modules.java", "(optional) org.jetbrains.java.decompiler".
+   * @return: list of plugin xml ids. Example: "org.jetbrains.kotlin
+   */
+  fun getPluginXmlIdByDependency(dependency: String, includeOptional: Boolean = true): List<String>
+
+  /**
+   * List of plugins compatible with [ideBuild]
+   */
+  fun listPlugins(ideBuild: String, channel: String? = null, pluginId: String? = null): List<PluginXmlBean>
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: I would not prefer to add an external library. let's discuss face to face

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 13]
```diff
@@ -1,21 +1,144 @@
...
+
+  /**
+   * List of plugins xml ids compatible with [build].
+   * @param max   - max result set. Max: 10000 - [offset]
+   * @param query - query for search.
+   */
+  fun getCompatiblePluginsXmlIds(build: String, max: Int, offset: Int, query: String = ""): List<String>
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: For this version, we don't need this param. 

>Also It's valid only for intellij plugins.

make sense

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (kesarevs) [reply README.md]: Like:

The code snippet below will init the instance of the JetBrains Marketplace repository. You need to provide a [permanent hub token](https://www.jetbrains.com/help/youtrack/standalone/Manage-Permanent-Token.html) to authorize.

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 14]
```diff
@@ -1,21 +1,144 @@
...
+   * @return list [CompatibleUpdateBean] if update exists. Get nothing if plugin is incompatible with [build]
+   */
+  fun searchCompatibleUpdates(xmlIds: List<String>, build: String, channel: String = ""): List<CompatibleUpdateBean>
+}
+
+interface PluginUpdateManager {
+  /**
+   * Get list of plugin updates by plugin [xmlId].
+   * @param version - version of the plugin.
+   * @return Normally method returns only ONE update, but several very OLD plugins have several update.
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: thanks

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: the old function name. will not change for this version

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 15]
```diff
@@ -1,21 +1,144 @@
...
+   * List of plugins xml ids compatible with [build].
+   * @param max   - max result set. Max: 10000 - [offset]
+   * @param query - query for search.
+   */
+  fun getCompatiblePluginsXmlIds(build: String, max: Int, offset: Int, query: String = ""): List<String>
+
+  /**
+   * Search last compatible update for each id from [xmlIds]
+   * @param channel - plugin channel. Default value is "stable" plugin channel.
+   * @return list [CompatibleUpdateBean] if update exists. Get nothing if plugin is incompatible with [build]
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: tnx

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 16]
```diff
@@ -1,21 +1,144 @@
...
-  fun uploadPlugin(pluginXmlId: String, file: File, channel: String? = null, notes: String? = null)
+  /**
+   * List of plugin supported [ProductEnum].
+   */
+  fun getPluginCompatibleProducts(id: Int): List<ProductEnum>
 
-  fun uploadNewPlugin(file: File, family: String, categoryId: Int, licenseUrl: String): PluginInfoBean
+  /**
+   * List of plugin dependencies. Examples: "com.intellij.modules.java", "(optional) org.jetbrains.java.decompiler".
+   * @return: list of plugin xml ids. Example: "org.jetbrains.kotlin
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: no. only xml ids

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply README.md]: will add. thx

[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply README.md]: thanks

**Comment by serejke** [comment]
Avoid using `@param <param name> - <description>` syntax. Separator `-` is useless and treated by Markdown wrongly as a list of items. Replace ` - ` with a single space. Check rendering using Quick Doc in IDEA.

**Comment by serejke** [comment]
Do not separate `json` \ `xml` model:
```
model/
   json/
   xml/
      converters/
```

It is an implementation detail whether a bean class comes from `json` or `xml`. I suggest moving all classes to `model` package and implementing proper conversions on `PluginRepositoryService` level.

**Comment by serejke** [comment]
`PluginBean` and `PluginXmlBean` are hardly distinguishable.

**Comment by serejke** [comment]
`PluginBean` and `PluginUpdateBean` are hardly distinguishable as well

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt [hunk 17]
```diff
@@ -0,0 +1,53 @@
+package org.jetbrains.intellij.pluginRepository.model.json
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import org.jetbrains.intellij.pluginRepository.model.repository.ProductFamily
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+data class PluginBean(
+  val id: Int,
+  val name: String,
+  val link: String,
```
Comment (serejke) [inline rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt]: This parameter looks useless given that there is `urls: PluginURLsBean` available

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+  val approve: Boolean,
```
Comment (serejke) [inline rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt]: This is an implementation detail. Non-approved plugins mustn't be included to plugin lists at all.

**Comment by serejke** [comment]
Please review nullable bean's parameters. It is unclear what `val removalRequested: Boolean?` when `removalRequested = null` means: is it `true` or `false` or `unspecified`. Anyway, Javadoc must clarify that.

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt (hunk 17 above)
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt]: you are right <3 I have just moved classes from the marketplace side.

Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt]: nope, link - its a link to plugin repository plugin page

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt (hunk 18 above)
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/model/json/PluginBean.kt]: you are right <3 I have just moved classes from the marketplace side.

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt [hunk 19]
```diff
@@ -1,21 +1,148 @@
...
+   * @param ideBuild - IDE version. Example: "IC-145.184"
+   * @param channel  - plugin channel. Default value is "stable" plugin channel.
+   */
+  fun downloadLatestCompatiblePlugin(xmlId: String, ideBuild: String, targetPath: File, channel: String? = null): File?
+}
+
+interface PluginUploader {
+  /**
+   * Upload plugin by ID into specific channel.
+   * IMPORTANT: PLUGIN NOTES WILL BE IGNORED FOR IDES BASED ON INTELLIJ PLATFORM ([ProductFamily.INTELLIJ]).
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: okay, thks =)

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (kesarevs) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: `Examples: "com.intellij.modules.java"` -- it's a module, I guess.

File: rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (kesarevs) [reply rest/src/main/kotlin/org/jetbrains/intellij/pluginRepository/PluginRepository.kt]: Then mark it as deprecated and create a new one with the correct name please.

File: cli/src/main/kotlin/org/jetbrains/intellij/pluginRepository/client.kt [hunk 20]
```diff
@@ -36,24 +37,27 @@ class Client {
...
         private fun upload(args: Array<String>) {
             val options = UploadOptions()
             Args.parseOrExit(options, args)
-            val pluginRepository = PluginRepositoryInstance(options.host, options.token)
+            val pluginRepository = PluginRepositoryFactory.create(
+              options.host, options.token).uploader
             val pluginId = options.pluginId
             when {
-                pluginId == null -> pluginRepository.uploadNewPlugin(File(options.pluginPath!!), options.family!!, 104, "https://plugins.jetbrains.com/legal/terms-of-use")
+                pluginId == null -> pluginRepository.uploadNewPlugin(File(options.pluginPath!!), 104,
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply cli/src/main/kotlin/org/jetbrains/intellij/pluginRepository/client.kt]: yes, some category id. not my changes

**Pull request merged**